File: kotlin/src/main/kotlin/algorithms/sorts.kt
package algorithms

fun MutableList<Int>.shakerSort() {

    var beginning = 0
    var end = this.size - 1

    do {
        var swapped = false
        for (i in beginning..<end) {
            if (this[i] > this[i + 1]) {
                val aux = this[i]
                this[i] = this[i + 1]
                this[i + 1] = aux
                swapped = true
            }
        }

        if (!swapped) break


        end--
        beginning++
        swapped = false

        for (i in end downTo  beginning) {
            if (this[i] < this[i - 1]) {
                val aux = this[i]
                this[i] = this[i - 1]
                this[i - 1] = aux
                swapped = true
            }
        }


    } while (swapped)
}

fun MutableList<Int>.bubbleSort() {

    var end = this.size - 1

    do {
        var swapped = false

        for (i in 0..<end) {
            if (this[i] > this[i + 1]) {
                val aux = this[i]
                this[i] = this[i + 1]
                this[i + 1] = aux
                swapped = true
            }
        }

        end--
    } while (swapped)
}

File: python/algorithms/sorts.py
from typing import List


def shaker_sort(vector: List[int]):
    swapped = True
    beginning = 0
    end = len(vector) - 1
    while swapped:
        swapped = False
        swapped = loop(vector, swapped, end)

        if not swapped:
            break

        swapped = False
        end -= 1


        for i in range(end, beginning, -1):
            if vector[i] < vector[i - 1]:
                vector[i], vector[i - 1] = vector[i - 1], vector[i]
                swapped = True

        beginning += 1

def bubble_sort(vector: List[int], swapped: bool = True):

    end = len(vector) - 1

    while swapped:
        swapped = False
        swapped = loop(vector, swapped, end)

        if not swapped:
            break

        end -= 1


def loop(vector: List[int], swapped: bool, end: int)-> bool:
    for i in range(end):
        if vector[i] > vector[i + 1]:
            vector[i], vector[i + 1] = vector[i + 1], vector[i]
            swapped = True

    return swapped